x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,359 @@
+    static final int MAX_PACKET_LENGTH_IN_BYTES = 1 << 20; // 1MB
+    static final int PACKET_LENGTH_IN_BYTES = 64 * (1 << 10); // 64KB
+
+    private static final String ENCRYPTION_METADATA_PREFIX = "encryption-metadata";
+
+    private final BlobStoreRepository delegatedRepository;
+    private final KeyGenerator dataEncryptionKeyGenerator;
+    private final PasswordBasedEncryptor metadataEncryptor;
+    private final ConsistentSettingsService consistentSettingsService;
+    private final SecureRandom secureRandom;
+    private final SecureSetting<?> passwordSettingForThisRepo;
+
+    protected EncryptedRepository(RepositoryMetaData metadata, NamedXContentRegistry namedXContentRegistry, ClusterService clusterService,
+                                  BlobStoreRepository delegatedRepository, PasswordBasedEncryptor metadataEncryptor,
+                                  ConsistentSettingsService consistentSettingsService) throws NoSuchAlgorithmException {
+        super(metadata, namedXContentRegistry, clusterService, delegatedRepository.basePath());
+        this.delegatedRepository = delegatedRepository;
+        this.dataEncryptionKeyGenerator = KeyGenerator.getInstance(EncryptedRepositoryPlugin.CIPHER_ALGO);
+        this.dataEncryptionKeyGenerator.init(DATA_KEY_SIZE_IN_BITS, SecureRandom.getInstance(EncryptedRepositoryPlugin.RAND_ALGO));
+        this.metadataEncryptor = metadataEncryptor;
+        this.consistentSettingsService = consistentSettingsService;
+        this.secureRandom = SecureRandom.getInstance(EncryptedRepositoryPlugin.RAND_ALGO);
+        this.passwordSettingForThisRepo =
+                (SecureSetting<?>) EncryptedRepositoryPlugin.ENCRYPTION_PASSWORD_SETTING.getConcreteSettingForNamespace(metadata.name());
+    }
+
+    @Override
+    public void snapshotShard(Store store, MapperService mapperService, SnapshotId snapshotId, IndexId indexId,
+                              IndexCommit snapshotIndexCommit, IndexShardSnapshotStatus snapshotStatus, boolean writeShardGens,
+                              ActionListener<String> listener) {
+        if (EncryptedRepositoryPlugin.getLicenseState().isEncryptedRepositoryAllowed()) {
+            if (consistentSettingsService.isConsistent(passwordSettingForThisRepo)) {
+                super.snapshotShard(store, mapperService, snapshotId, indexId, snapshotIndexCommit, snapshotStatus, writeShardGens, listener);
+            } else {
+                listener.onFailure(new RepositoryException(metadata.name(),
+                        "Password mismatch for repository [" + metadata.name() + "]. The local node's value of the " +
+                                "keystore secure setting [" + passwordSettingForThisRepo.getKey() + "] is different from the master's"));
+            }
+        } else {
+            listener.onFailure(LicenseUtils.newComplianceException(
+                    EncryptedRepositoryPlugin.REPOSITORY_TYPE_NAME + " snapshot repository"));
```

Comment:
@tvernum @jkakavas We discussed and agreed that we require the platinum license for encrypted repositories (I lost the link to the discussion). But I think it makes sense to require the license only for the snapshotting operation. To me it feels like a rip-off to require a license to retrieve the data which has been backed up when a license was in place.

Do we agree that only the snapshot operation falls under the license requirement? If so, this is the best I could come up with. This is again less than ideal because the snapshot operation does not fail as soon as possible. Here is how the response for an invalid license looks like:
```
{
  "snapshot" : {
    "snapshot" : "snapshot_6",
    "uuid" : "1NTz-Tm7SWeIQBGegbvROQ",
    "version_id" : 8000099,
    "version" : "8.0.0",
    "indices" : [
      "twitter2",
      "twitter"
    ],
    "include_global_state" : true,
    "state" : "PARTIAL",
    "start_time" : "2020-01-15T13:51:02.234Z",
    "start_time_in_millis" : 1579096262234,
    "end_time" : "2020-01-15T13:51:02.234Z",
    "end_time_in_millis" : 1579096262234,
    "duration_in_millis" : 0,
    "failures" : [
      {
        "index" : "twitter",
        "index_uuid" : "twitter",
        "shard_id" : 0,
        "reason" : "org.elasticsearch.ElasticsearchSecurityException: current license is non-compliant for [encrypted snapshot repository]\n\tat org.elasticsearch.license.LicenseUtils.newComplianceException(LicenseUtils.java:26)\n\tat org.elasticsearch.repositories.encrypted.EncryptedRepository.snapshotShard(EncryptedRepository.java:104)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.snapshot(SnapshotShardsService.java:341)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.lambda$startNewShards$1(SnapshotShardsService.java:287)\n\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:629)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:830)\n",
        "node_id" : "tA8IRz40TAq5gJpo3i4paA",
        "status" : "INTERNAL_SERVER_ERROR"
      },
      {
        "index" : "twitter2",
        "index_uuid" : "twitter2",
        "shard_id" : 1,
        "reason" : "org.elasticsearch.ElasticsearchSecurityException: current license is non-compliant for [encrypted snapshot repository]\n\tat org.elasticsearch.license.LicenseUtils.newComplianceException(LicenseUtils.java:26)\n\tat org.elasticsearch.repositories.encrypted.EncryptedRepository.snapshotShard(EncryptedRepository.java:104)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.snapshot(SnapshotShardsService.java:341)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.lambda$startNewShards$1(SnapshotShardsService.java:287)\n\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:629)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:830)\n",
        "node_id" : "tA8IRz40TAq5gJpo3i4paA",
        "status" : "INTERNAL_SERVER_ERROR"
      },
      {
        "index" : "twitter2",
        "index_uuid" : "twitter2",
        "shard_id" : 0,
        "reason" : "org.elasticsearch.ElasticsearchSecurityException: current license is non-compliant for [encrypted snapshot repository]\n\tat org.elasticsearch.license.LicenseUtils.newComplianceException(LicenseUtils.java:26)\n\tat org.elasticsearch.repositories.encrypted.EncryptedRepository.snapshotShard(EncryptedRepository.java:104)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.snapshot(SnapshotShardsService.java:341)\n\tat org.elasticsearch.snapshots.SnapshotShardsService.lambda$startNewShards$1(SnapshotShardsService.java:287)\n\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:629)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:830)\n",
        "node_id" : "tA8IRz40TAq5gJpo3i4paA",
        "status" : "INTERNAL_SERVER_ERROR"
      }
    ],
    "shards" : {
      "total" : 3,
      "failed" : 3,
      "successful" : 0
    }
  }
}
```

I think this is reasonable, given the hooks we currently have in place. The error and the reason for the user are clear, although it does create a "snapshot stub" , i.e. a snapshot only for the global metadata. What do you think?

Reply:
I've learned that the `Repository#getRepositoryData` method is always called on the master node before the request fans out. The license check is better suited there. I will move the check and return back with the new user experience.

Reply:
I have an idea of making the repository `read-only` if the license is non-platinum (and non-trial). I think this would require a little bit of code shuffling, because the `read-only` flag is assumed to not change, but I believe it's feasible. What do you think @original-brownbear ? Making the repo `read-only` precludes snapshot deletes and clean-ups, which is unfortunate but it's not terrible. Also, it might take a while to learn why the repository flipped to `read-only`, although you'd probably hit other surprises first on an expired platinum license.

 Otherwise, I would propose we add a new method to the `Repository` class `isSnapshotEnabled()`; I envisage there might be other future cases where only partial operations on the repo are enabled but the conditions for their availability can externally change.

Reply:
`read-only` isn't enough to make restores fail-fast though. I guess we could enable all other read operations somehow so you would still have the snapshot status APIs but it would require some API changes and I'm wondering if the limited usefulness of that is worth it?

Reply:
My assumption, but I have to get @tvernum and @jkakavas input, is that restore IS allowed for a `non-platinum` license, but snapshot IS NOT (and delete and clean-up are collateral). We could simply disable the repository completely in a case of non-conformant license (failing the `Repositry#getRepositoryData` call) but to me this sounds like racketeering: if you at some point have created a snapshot, then you must pay us whenever in the future you need to restore it.

Reply:
You still can get a trial or hack the license check code, so there won't be any tears, but still it sounds dishonest.

Reply:
I see nothing dishonest with tying a feature ( " if you at some point have created a snapshot, " --> " if you at some point have created an _encrypted_ snapshot, " ) to a license level. Maybe worth taking in the view of @bytebilly from a PM perspective.

Reply:
Discussed in https://github.com/elastic/dev/issues/1258